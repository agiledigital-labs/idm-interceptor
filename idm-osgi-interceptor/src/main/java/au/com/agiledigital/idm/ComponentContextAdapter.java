package au.com.agiledigital.idm;

import java.util.Dictionary;
import java.util.Hashtable;

import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;
import org.osgi.service.component.ComponentContext;
import org.osgi.service.component.ComponentInstance;

public class ComponentContextAdapter implements ComponentContext {
	
	BundleContext bundleContext;
	Dictionary<String, Object> properties;

	public ComponentContextAdapter(ServiceReference<?> ref) {
		super();
		this.bundleContext = ref.getBundle().getBundleContext();
		
		Hashtable<String, Object> properties = new Hashtable<String, Object>();
		for (String key : ref.getPropertyKeys()) {
			properties.put(key, ref.getProperty(key));
		}
		this.properties = properties;
	}

	@Override
	public void disableComponent(String arg0) {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public void enableComponent(String arg0) {
		throw new IllegalStateException("Not implemented");	}

	@Override
	public BundleContext getBundleContext() {
		return bundleContext;
	}

	@Override
	public ComponentInstance getComponentInstance() {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public Dictionary<String, Object> getProperties() {
		return this.properties;
	}

	@Override
	public ServiceReference<?> getServiceReference() {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public Bundle getUsingBundle() {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public Object locateService(String arg0) {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public <S> S locateService(String arg0, ServiceReference<S> arg1) {
		throw new IllegalStateException("Not implemented");
	}

	@Override
	public Object[] locateServices(String arg0) {
		throw new IllegalStateException("Not implemented");
	}
}
