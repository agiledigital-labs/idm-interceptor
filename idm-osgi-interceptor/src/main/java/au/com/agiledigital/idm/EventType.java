package au.com.agiledigital.idm;

public enum EventType {
	CREATE("create"),
	UPDATE("update"),
	DELETE("delete"),
	READ("read"),
	ACTION("action"),
	PATCH("patch"),
	QUERY("query");

	private String name;
	EventType(String name) {
		this.name = name;
	}
	
	public String getName() {
		return this.name;
	}
}
