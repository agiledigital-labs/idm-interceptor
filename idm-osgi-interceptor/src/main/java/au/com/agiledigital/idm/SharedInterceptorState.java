package au.com.agiledigital.idm;

import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import org.apache.felix.dm.annotation.api.Component;
import org.apache.felix.dm.annotation.api.ServiceScope;
import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.ObjectClass;
import org.identityconnectors.framework.common.objects.Uid;

import au.com.agiledigital.idm.connector.api.DummyConnectorApi;

@Component(scope = ServiceScope.SINGLETON)
public class SharedInterceptorState implements SharedInterceptorStateService {

	private ConcurrentHashMap<String, ConcurrentMap<DummyConnectorApi, Boolean>> connectorMap = new ConcurrentHashMap<>();
	private ConcurrentHashMap<String, ConcurrentMap<ObjectClass, ConcurrentMap<Uid, Set<Attribute>>>> connectorData = new ConcurrentHashMap<>();

	private String eventEndpoint;

	@Override
	public ConcurrentMap<ObjectClass, ConcurrentMap<Uid, Set<Attribute>>> registerConnector(String factoryPid,
			DummyConnectorApi dummyConnector) {
		connectorMap.computeIfAbsent(factoryPid, key -> new ConcurrentHashMap<>()).put(dummyConnector, true);

		// We create the connector data map here so that state isn't lost when using a
		// poolable connector across multiple threads.
		return connectorData.computeIfAbsent(factoryPid, key -> new ConcurrentHashMap<>());
	}

	@Override
	public void unregisterConnector(String factoryPid, DummyConnectorApi dummyConnector) {
		connectorMap.get(factoryPid).remove(dummyConnector);
	}

	@Override
	public void setEventEndpoint(String factoryPid) {
		this.eventEndpoint = factoryPid;
	}

	@Override
	public Optional<String> getEventEndpoint() {
		return Optional.ofNullable(this.eventEndpoint);
	}

	@Override
	public Map<String, String> clearAllConnectorData() {
		Map<String, String> results = new ConcurrentHashMap<>();
		
		// Because the connectors use shared state, where we have a pool of connectors, we can simply use the first in the set. 
		this.connectorMap.entrySet().parallelStream().filter(entry -> entry.getValue().size() > 0)
				.forEach(entry -> results.put(entry.getKey(), entry.getValue().keySet().iterator().next().clearData()));
		return results;
	}
}
