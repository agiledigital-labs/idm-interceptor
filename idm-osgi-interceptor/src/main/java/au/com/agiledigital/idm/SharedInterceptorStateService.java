package au.com.agiledigital.idm;

import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentMap;

import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.ObjectClass;
import org.identityconnectors.framework.common.objects.Uid;

import au.com.agiledigital.idm.connector.api.DummyConnectorApi;

public interface SharedInterceptorStateService {

	ConcurrentMap<ObjectClass, ConcurrentMap<Uid, Set<Attribute>>> registerConnector(String factoryPid, DummyConnectorApi dummyConnector);

	void unregisterConnector(String factoryPid, DummyConnectorApi dummyConnector);

	void setEventEndpoint(String factoryPid);

	Optional<String> getEventEndpoint();

	Map<String, String> clearAllConnectorData();
}