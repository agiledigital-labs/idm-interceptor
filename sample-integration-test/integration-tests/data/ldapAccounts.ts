export const initialValues = [
    {
        "_id": "existing-bjensen",
        "dn": "uid=bjensen,ou=people,dc=agiledigital,dc=com,dc=au",
        "mail": "bjensen@example.com",
        "telephoneNumber": "1234567",
        "givenName": "Barbara",
        "uid": "bjensen",
        "sn": "Jensen",
        "description": "Created By LDAP"
    },
    {
        "_id": "existing-scarter",
        "dn": "uid=scarter,ou=people,dc=agiledigital,dc=com,dc=au",
        "mail": "scarter@example.com",
        "telephoneNumber": "1234567",
        "givenName": "Steven",
        "uid": "scarter",
        "sn": "Carter",
        "description": "Created By LDAP"
    }
];