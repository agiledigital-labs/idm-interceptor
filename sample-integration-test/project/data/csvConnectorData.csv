"description","uid","username","firstname","lastname","email","mobileTelephoneNumber","roles"
"Created By CSV","bjensen","bjensen@example.com","Barbara","Jensen","bjensen@example.com","1234567","internal/role/openidm-authorized"
"Created By CSV","scarter","scarter@example.com","Steven","Carter","scarter@example.com","1234567","internal/role/openidm-admin,internal/role/openidm-authorized"
