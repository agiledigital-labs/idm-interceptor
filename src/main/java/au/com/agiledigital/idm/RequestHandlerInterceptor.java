package au.com.agiledigital.idm;

import java.util.Map;
import java.util.Set;

import org.apache.felix.dm.annotation.api.AspectService;
import org.apache.felix.dm.annotation.api.ServiceDependency;
import org.forgerock.json.JsonValue;
import org.forgerock.json.resource.ActionRequest;
import org.forgerock.json.resource.ActionResponse;
import org.forgerock.json.resource.CreateRequest;
import org.forgerock.json.resource.DeleteRequest;
import org.forgerock.json.resource.PatchRequest;
import org.forgerock.json.resource.QueryRequest;
import org.forgerock.json.resource.QueryResourceHandler;
import org.forgerock.json.resource.QueryResponse;
import org.forgerock.json.resource.ReadRequest;
import org.forgerock.json.resource.RequestHandler;
import org.forgerock.json.resource.ResourceException;
import org.forgerock.json.resource.ResourcePath;
import org.forgerock.json.resource.ResourceResponse;
import org.forgerock.json.resource.UpdateRequest;
import org.forgerock.openidm.config.enhanced.EnhancedConfig;
import org.forgerock.openidm.crypto.CryptoService;
import org.forgerock.openidm.provisioner.openicf.commons.ConnectorUtil;
import org.forgerock.openidm.provisioner.openicf.commons.ObjectClassInfoHelper;
import org.forgerock.services.context.Context;
import org.forgerock.util.promise.Promise;
import org.identityconnectors.framework.common.objects.Attribute;
import org.osgi.framework.ServiceReference;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@AspectService(ranking = 10, added = "added", changed = "changed", removed = "removed", filter = "(&(objectClass=org.forgerock.json.resource.RequestHandler)(service.factoryPid=org.forgerock.openidm.provisioner.openicf))")
public class RequestHandlerInterceptor implements RequestHandler {
	private volatile RequestHandler intercepted;
	private volatile String factoryPid;
	@ServiceDependency
	private volatile EnhancedConfig enhancedConfig;
	
	@ServiceDependency
	private CryptoService cryptoService;
	
	private Map<String, ObjectClassInfoHelper> objectClassInfoMap;

	private static Logger logger = LoggerFactory.getLogger(RequestHandlerInterceptor.class);
	
	private Map<String, ObjectClassInfoHelper> generateObjectClassInfo(ServiceReference<?> ref) {
		JsonValue jsonConfig = this.enhancedConfig.getConfigurationAsJson(new ComponentContextAdapter(ref));
		Map<String, ObjectClassInfoHelper> objectTypes = ConnectorUtil.getObjectTypes(jsonConfig);
		
		return objectTypes;
	}
	
	private ObjectClassInfoHelper getObjectClassInfoHelper(ResourcePath resourcePath) {
		return objectClassInfoMap.get(resourcePath.get(0));
	}

	public void added(ServiceReference<RequestHandler> ref, RequestHandler service) {
		String factoryPid = (String) ref.getProperty("config.factory-pid");
		this.objectClassInfoMap = generateObjectClassInfo(ref);
		this.factoryPid = factoryPid;
		this.intercepted = service;

		logger.info("Added Service {}", factoryPid);
	}

	public void changed(ServiceReference<RequestHandler> ref, RequestHandler service) {
		String factoryPid = (String) ref.getProperty("config.factory-pid");
		this.objectClassInfoMap = generateObjectClassInfo(ref);
		this.factoryPid = factoryPid;
		this.intercepted = service;
		
		logger.info("Changed Service {}", factoryPid);
	}
	
	public void removed(ServiceReference<RequestHandler> ref, RequestHandler service) {
		String factoryPid = (String) ref.getProperty("config.factory-pid");
		this.intercepted = null;
		this.factoryPid = null;
		this.objectClassInfoMap = null;
		
		logger.info("Removed Service {}", factoryPid);
	}
	
	private void ensureHasValue() {
		if (intercepted == null || factoryPid == null) {
			throw new IllegalStateException("The intercepted instance is missing, cannot proceed");
		}
	}

	@Override
	public Promise<ActionResponse, ResourceException> handleAction(Context context, ActionRequest request) {
		ensureHasValue();
		logger.info("About to execute action for {}", factoryPid);
		return intercepted.handleAction(context, request);
	}

	@Override
	public Promise<ResourceResponse, ResourceException> handleRead(Context context, ReadRequest request) {
		ensureHasValue();
		logger.info("About to read for {}", factoryPid);
		Promise<ResourceResponse, ResourceException> handleRead = intercepted.handleRead(context, request);
		return handleRead;
	}
	
	@Override
	public Promise<ResourceResponse, ResourceException> handleCreate(Context context, CreateRequest request) {
		ensureHasValue();
		logger.info("About to create for {}", factoryPid);
		
		ObjectClassInfoHelper helper = getObjectClassInfoHelper(request.getResourcePathObject());
		
		try {
			Set<Attribute> createAttributes = helper.getCreateAttributes(request, this.cryptoService);
			logger.info("Create attributes {}", createAttributes);
		} catch (ResourceException e) {
			return e.asPromise();
		}
		
		return intercepted.handleCreate(context, request);
	}
	
	@Override
	public Promise<ResourceResponse, ResourceException> handleDelete(Context context, DeleteRequest request) {
		ensureHasValue();
		logger.info("About to delete for {}", factoryPid);
		return intercepted.handleDelete(context, request);
	}
	
	@Override
	public Promise<ResourceResponse, ResourceException> handlePatch(Context context, PatchRequest request) {
		ensureHasValue();
		logger.info("About to Patch for {}", factoryPid);
		return intercepted.handlePatch(context, request);
	}
	
	@Override
	public Promise<QueryResponse, ResourceException> handleQuery(Context context, QueryRequest request,
			QueryResourceHandler handler) {
		ensureHasValue();
		logger.info("About to Query for {}", factoryPid);
		return intercepted.handleQuery(context, request, handler);
	}
	
	@Override
	public Promise<ResourceResponse, ResourceException> handleUpdate(Context context, UpdateRequest request) {
		ensureHasValue();
		logger.info("About to Update for {}", factoryPid);
		return intercepted.handleUpdate(context, request);
	}
}
