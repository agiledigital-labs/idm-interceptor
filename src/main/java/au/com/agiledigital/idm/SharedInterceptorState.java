package au.com.agiledigital.idm;

import java.util.concurrent.ConcurrentHashMap;

import org.apache.felix.dm.annotation.api.Component;
import org.apache.felix.dm.annotation.api.ServiceScope;

import au.com.agiledigital.idm.interceptor.RequestHandlerInterceptor;

@Component(scope = ServiceScope.SINGLETON)
public class SharedInterceptorState implements SharedInterceptorStateService {
	
	private ConcurrentHashMap<String, RequestHandlerInterceptor> interceptorMap = new ConcurrentHashMap<>();
	
	@Override
	public void registerInterceptor(String factoryPid, RequestHandlerInterceptor interceptor) {
		interceptorMap.put(factoryPid, interceptor);
	}
	
	@Override
	public void unregisterInterceptor(String factoryPid, RequestHandlerInterceptor interceptor) {
		interceptorMap.remove(factoryPid);
	}
}
