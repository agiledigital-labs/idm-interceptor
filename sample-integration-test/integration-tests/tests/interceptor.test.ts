import openidmLib from '../util/openidm-lib';
import each from 'jest-each';
import idm from 'lib/idm';
import { equals, startsWith, trueVal } from '@agiledigital/idm-ts-types/lib/query-filter';
import * as ldapAccounts from '../data/ldapAccounts';

export const clearAll = () => {
    each`
        name                  | resource
        ${"Users"}            | ${openidmLib.managed.users}
        ${"Links"}            | ${openidmLib.links}
    `.test('$name should be reset to be empty', async ( { resource} ) => {
        await resource.deleteAll();
        const result =  await resource.getAll().expect(200)
        expect(result.body.resultCount).toBe(0);
    }, 15000)

    test('Clear Dummy Data and Connector Events', async() => {
        const status = await openidmLib.connectorEvents.clearAll().then(res => res.status);
        expect(status).toBe(200);
    }, 1000);
};

describe('No data in LDAP, creates are performed', () => {
    clearAll();
    test('reconcile csv', async () => {
        const status = await openidmLib.reconcile.csvToManagedUser().then(res => res.body);
        expect(status.state).toBe('SUCCESS');
    }, 3000);
    test('created in LDAP', async () => {
        // There were 2 create events
        const ldapCreateEvents = await idm.system.connectoreventsCreate.query({filter: equals("objectType", "system/adeldap/account")});
        expect(ldapCreateEvents.resultCount).toBe(2);

        // And no update events
        const ldapUpdateEvents = await idm.system.connectoreventsUpdate.query({filter: equals("objectType", "system/adeldap/account")});
        expect(ldapUpdateEvents.resultCount).toBe(0);

        const adeLdapQuery = await idm.system.adeldapAccount.query({filter: trueVal()});

        adeLdapQuery.result.forEach(acc => {
            expect(acc).toMatchSnapshot({
                _id: expect.any(String)
            }, `Expected LDAP Created Account [${acc.uid}]`)
        });
    }, 5000);
});

describe('Existing data in LDAP, updates are performed', () => {
    clearAll();
    each`
        name                    | values                                  | resource
        ${"LDAP accounts"}      | ${ldapAccounts.initialValues}           | ${openidmLib.system.ldap.accounts}
    `.test('loading $name data', async ( { values, resource }) => {
        const result = await resource.createAll(values);
        const failures = result.filter(item => item.status !== 201);
        expect(failures.length).toBe(0);
    });
    test('check for create events', async () => {
        // There were 2 create events
        const ldapCreateEvents = await idm.system.connectoreventsCreate.query({filter: equals("objectType", "system/adeldap/account")});
        expect(ldapCreateEvents.resultCount).toBe(2);
    
        // Clear the events before reconciling
        const statusClearEvents = await openidmLib.connectorEvents.clearEvents().then(res => res.status);
        expect(statusClearEvents).toBe(200);
    }, 3000);
    test('reconcile csv', async () => {
        const status = await openidmLib.reconcile.csvToManagedUser().then(res => res.body);
        expect(status.state).toBe('SUCCESS');
    }, 3000);
    test('updated in LDAP', async () => {
        // There were 0 create events
        const ldapCreateEvents = await idm.system.connectoreventsCreate.query({filter: equals("objectType", "system/adeldap/account")});
        expect(ldapCreateEvents.resultCount).toBe(0);

        // And 2 update events
        const ldapUpdateEvents = await idm.system.connectoreventsUpdate.query({filter: startsWith("objectType", "system/adeldap/account")});
        expect(ldapUpdateEvents.resultCount).toBe(2);

        const adeLdapQuery = await idm.system.adeldapAccount.query({filter: trueVal()});

        adeLdapQuery.result.forEach(acc => {
            expect(acc).toMatchSnapshot({
                _id: expect.any(String)
            }, `Expected LDAP Updated Account [${acc.uid}]`)
        });
    }, 5000);
});