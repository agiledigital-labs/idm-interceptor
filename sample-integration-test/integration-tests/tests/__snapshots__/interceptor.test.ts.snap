// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Existing data in LDAP, updates are performed updated in LDAP: Expected LDAP Updated Account [bjensen] 1`] = `
Object {
  "_id": Any<String>,
  "description": "Created By CSV",
  "dn": "uid=bjensen,ou=people,dc=agiledigital,dc=com,dc=au",
  "givenName": "Barbara",
  "mail": "bjensen@example.com",
  "sn": "Jensen",
  "telephoneNumber": "1234567",
  "uid": "bjensen",
}
`;

exports[`Existing data in LDAP, updates are performed updated in LDAP: Expected LDAP Updated Account [scarter] 1`] = `
Object {
  "_id": Any<String>,
  "description": "Created By CSV",
  "dn": "uid=scarter,ou=people,dc=agiledigital,dc=com,dc=au",
  "givenName": "Steven",
  "mail": "scarter@example.com",
  "sn": "Carter",
  "telephoneNumber": "1234567",
  "uid": "scarter",
}
`;

exports[`No data in LDAP, creates are performed created in LDAP: Expected LDAP Created Account [bjensen] 1`] = `
Object {
  "_id": Any<String>,
  "description": "Created By CSV",
  "dn": "uid=bjensen,ou=people,dc=agiledigital,dc=com,dc=au",
  "givenName": "Barbara",
  "mail": "bjensen@example.com",
  "sn": "Jensen",
  "telephoneNumber": "1234567",
  "uid": "bjensen",
}
`;

exports[`No data in LDAP, creates are performed created in LDAP: Expected LDAP Created Account [scarter] 1`] = `
Object {
  "_id": Any<String>,
  "description": "Created By CSV",
  "dn": "uid=scarter,ou=people,dc=agiledigital,dc=com,dc=au",
  "givenName": "Steven",
  "mail": "scarter@example.com",
  "sn": "Carter",
  "telephoneNumber": "1234567",
  "uid": "scarter",
}
`;
